### ISSUE TITLE
Spurious error logged prior to devices being listed

### ISSUE THREAD

**Issue by richnew10** [aws/amazon-chime-sdk-js#1568]
Stack:

```
getGroupIdFromDeviceId (DefaultDeviceController.js?66c1:1016)
chooseInputIntrinsicDevice (DefaultDeviceController.js?66c1:1131)
acquireInputStream (DefaultDeviceController.js?66c1:1376)
acquireAudioInputStream (DefaultDeviceController.js?66c1:585)
run (ReceiveAudioInputTask.js?ae4c:26)
```

Choosing a device requires computing the group ID to populate the `DeviceSelection`. For some reason we do so by consulting the device list cache. If the cache doesn't exist because no devices have been listed, any attempt to select a device — even a stream or the null device — will cause this to be hit:

```typescript
    } else {
      this.logger.error(
        'Device cache is not populated. Please make sure to call list devices first'
      );
```

This was introduced in #1329.

Instead we should:

* Not use a group ID for the null device.
* Use the group ID from the settings of a media stream.
* Consider enumerating devices as needed in order to do the work.
* Probably not log an error, because applications seem to continue to work without this.

I have started to look at this, but will hand it off if I can't get it wrapped up.

**Comment by ltrung** [comment]
@richnew10 I think we assume you always need to call `list<Audio/Video>InputDevice` first before calling the corresponding choose API and the current code would not work if device cache is not populated due to this check https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L1388. This whole complication with device cache comes from that logic. In some browsers (I think Chrome), deviceId can be the same (`default`) so we need to rely on groupId to know if they are the same device.  However, since the input info does not contain groupId, we have to rely on the device cache. 
Also if this check fails, it will cause infinite cycle with `chooseAudioInputDevice` since that API will always call `chooseInputIntrinsicDevice` twice (https://github.com/aws/amazon-chime-sdk-js/blame/66cd9809ea156a66dbd988c866fef9b489ca7c20/src/devicecontroller/DefaultDeviceController.ts#L1747). 
Therefore, at least in the case with audio input, if you see that error message, you will probably see infinite cycle also and application would not work.

We do store groupId from settings of the active media stream in the `activeDevices` object here: https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L1491. However, in some odd case (from another issues opened by customers), the groupId returned from the browsers are not in the list of enumerateDevices. So #1410 has to make sure that groupId exists in device cache. 

Also I believe we set groupId for null device to ''. 

I do agree that maybe we could do better to just enumerating devices as needed in order to do the work. At that time, we just added logs to help with debug multiple reported GitHub issues and we think since in all of our code examples, we ask to run list before choose, its probably fine to error out so builders know the need to call list.

A bit long reply but just want to give you a full context about this. Feel free to ping me if you want to discuss as I work a fair bit on this code :)

**Comment by richnew10** [comment]
Yes, exactly that — but WebRTC applications in general _do not_ expect to have to do so. Best practice is to hold on to device IDs and reuse them across sessions, only enumerating devices if needed; it's probably not right to force them to get another mic input and enumerate devices before picking the one they want.

This is definitely a complex chunk of code.

We have one large customer that routinely sees that stack trace, and their app is working.

**Comment by ltrung** [comment]
Ideally yes, we should not do this but browsers do not work the same way and Chrome relies on groupId :)

**Comment by richnew10** [comment]
Fixed by #1577.
